code/framework/auditing/src/main/java/io/cattle/platform/iaas/api/auditing/AuditServiceImpl.java

```diff
@@ -0,0 +1,160 @@
+        String resourceType = request.getType();
+        String eventType = resourceType + "." + (requestObject.get("action") != null ? (String) requestObject.get
+                ("action") :convertToAction(request.getMethod()));
+        publishEvent(auditLogDao.create(resourceType, parseId(resourceId), data, user,
+                policy.getAccountId(), policy.getAuthenticatedAsAccountId(), eventType, authType, runtime, null));
+    }
+
+    private Long parseId(String resourceId) {
+        Long parsedResourceId;
+        if (resourceId == null){
+            parsedResourceId = null;
+        } else {
+            try {
+                parsedResourceId = Long.valueOf(ApiContext.getContext().getIdFormatter().parseId(resourceId));
+            } catch (NumberFormatException e){
+                parsedResourceId = null;
+            }
+        }
+        return parsedResourceId;
+    }
+
+    private AuditEventType convertToAction(String method) {
+        switch (Schema.Method.valueOf(method)){
+            case DELETE:
+                return AuditEventType.delete;
+            case POST:
+                return AuditEventType.create;
+            case PUT:
+                return AuditEventType.update;
+            default:
+                return AuditEventType.UNKNOWN;
+        }
+    }
+
+    @Override
+    public void logResourceModification(Object resource, Map<String, Object> data, AuditEventType eventType,
+                                        String description, Long accountId) {
+        putInAsString(data, "resource", "Failed to convert resource to json.", resource);
+        publishEvent(auditLogDao.create(objectManager.getType(resource), parseId((String) ObjectUtils.getId(resource)),
+                data, null, accountId, null, eventType.toString(), null, null, description));
+    }
+
+    protected void publishEvent(AuditLog auditLog) {
+        Map<String, Object> data = new HashMap<>();
+        data.put(ObjectMetaDataManager.ACCOUNT_FIELD, auditLog.getAccountId());
+
+        Event event = EventVO.newEvent(FrameworkEvents.STATE_CHANGE)
+                .withData(data)
+                .withResourceType(auditLog.getClass().getName())
+                .withResourceId(auditLog.getId().toString());
+
+        eventService.publish(event);
+    }
+
+    private void putInAsString(Map<String, Object> data, String fieldForObject, String errMsg, Object objectToPlace) {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        try {
+            jsonMapper.writeValue(os, objectToPlace);
+            data.put(fieldForObject, os.toString());
+        } catch (IOException e) {
+            log.error(errMsg);
+            data.put("requestObject", "{\"object\":\"" + errMsg + "\"");
+        }
+    }
+
+    private boolean containsIgnoreCase(Set<String> list, String string){
+        for(String i : list){
```

Comment:
just lower case everything.  so have the set be all lowercase and then lower case the value you are checking.  then its `O(1)`